source-code/core/plugin/src/parsePlugin.ts

```diff
@@ -0,0 +1,86 @@
+import type { InlangConfig } from "@inlang/config"
+import { tryCatch } from "@inlang/result"
+import { PluginApi, pluginIdRegex } from "./api.js"
+import {
+	PluginIncorrectlyDefinedUsedApisError,
+	PluginUsesReservedNamespaceError,
+	PluginUsesUnavailableApiError,
+	PluginInvalidIdError,
+	PluginApiAlreadyDefinedError,
+} from "./errors.js"
+import type { ResolvePluginResult } from "./resolvePlugins.js"
+
+/**
+ * Plugin ids that should be excluded by namespace reserving.
+ */
+const whitelistedPlugins = [
+	"inlang.plugin-json",
+	"inlang.plugin-standard-lint-rules",
+	"inlang.plugin-i18next",
+]
+
+export const parsePlugin = (args: {
+	config: InlangConfig
+	plugin: PluginApi
+	pluginInConfig: InlangConfig["plugins"][0]
+	result: ResolvePluginResult
+}) => {
```

Comment:
@felixhaeberle I would not mix parsing with logic. 

```ts
function parsePlugin(maybeValidPlugin: Plugin): Result<Plugin, Error> {}
```

Hiding callbacks behind a setup function makes things complicated. Maybe we should go for the alternative `setup` pattern that we discussed: 

```ts
type Plugin = { 
  meta: Metadata,
  // called on resolving the plugin. can be used to initialize state/whatever. 
  setup: (args: { options: unknown } ) => void
  loadMessages(),
  saveMessages(),
  addAppSpecificApis(),
  // ...
  
}
```

@felixhaeberle what do you think?

Reply:
Also makes `meta.usedApis` redundant because we know what APIs are used. In the future, we can re-add a meta.usedApis if we need to serialize the data.

Reply:
yep, we should improve a few things there structure-wise.

Reply:
> (because of multiple plugins in conjunction causing errors)

You are mixing parsing errors and resolution errors. 

`parsePlugin` = is a plugin valid
`resolvePlugins` = are all plugins in combination valid

Reply:
oh, yes. we need to restructure the code there.